File: queue/queue.ts
export class Queue<T> {
  private queue: T[] = [];
  private head: number;
  private tail: number;
  size: number;

  constructor(size: number) {
    this.queue = new Array(size)
    this.head = 0;
    this.tail = 0;
    this.size = size;
  }

  expand() {
    this.size = this.size * 2
  }

  enqueue(item: T) {
    if (this.tail === this.size) {
      if (this.head > 0) {
        for (let i = 0; i < this.size; i++) {
          this.queue[i] = this.queue[this.head + i]
        }
        this.tail = this.size - this.head;
        this.head = 0;
      } else {
        return false
      }
    }

    this.queue[this.tail] = item;
    this.tail += 1;
    return true
  }

  dequeue() {
    if (this.head === this.tail) {
      return null
    }
    
    return this.queue[this.head++];
  }
}

export class LoopQueue<T> {
  private queue: T[] = [];
  private head: number;
  private tail: number;
  size: number;
  count = 0;

  constructor(size: number) {
    this.queue = new Array(size)
    this.head = 0;
    this.tail = 0;
    this.size = size;
  }

  moveIndex(idx: number) {
    return (idx + 1) % this.size
  }

  enqueue(item: T) {
    if (this.count === this.size) {
      return false
    } else {
      this.queue[this.tail] = item;
      this.tail = this.moveIndex(this.tail)
      this.count++;
    }
  }

  dequeue() {
    if (this.count > 0) {
      const item = this.queue[this.head];
      this.head = this.moveIndex(this.head)
      this.count--;
      return item
    }
    return null
  }
}

const queue = new LoopQueue<number>(3);

File: queue/queue.rs
#[derive(Debug)]
struct Queue<T> {
    queue: Vec<T>,
    head: usize,
    tail: usize,
}

impl<T: Clone> Queue<T> {

    fn new(size: usize) -> Self {
        Queue {
            queue: Vec::with_capacity(size),
            head: 0,
            tail: 0,
        }
    }

    fn enququ(&mut self, item: T) -> bool {
        let c = self.queue.capacity();
        if self.head == 0 && self.tail == c {
            return false
        }

        if self.head > 0 {
            for i in 0..self.head {
                self.queue[i] = self.queue[self.head + i].clone();
            }
            self.tail = self.tail - self.head;
            self.head = 0;

            self.queue[self.tail] = item;
        } else {
            self.queue.push(item);
        }
        self.tail += 1;
        true
    }

    fn dequeue(&mut self) -> Option<T> {
        if self.head < self.tail {
            let item = self.queue[self.head].clone();
            self.head += 1;
            Some(item)
        } else {
            None
        }
    }

    fn expand(&mut self) {
        let old_capacity = self.queue.capacity();
        let ne_capacity = old_capacity * 2;
        let mut new_queue = Vec::with_capacity(ne_capacity);
        for i in self.head..self.tail {
            new_queue.push(self.queue[i].clone());
        }
        self.queue = new_queue;
        self.tail = self.tail - self.head;
        self.head = 0;
    }
}